File: lib/services/dockerService.ts
/**
 * Docker Database Service
 * 
 * TypeScript service for managing Docker database containers
 * via CUBE Nexum backend (Bollard/Docker API)
 */

import { invoke } from '@tauri-apps/api/core';

// ============================================================================
// Types
// ============================================================================

export type DatabaseType = 'postgresql' | 'mysql' | 'mongodb' | 'redis';

export type ContainerStatus =
  | 'created'
  | 'running'
  | 'paused'
  | 'restarting'
  | 'removing'
  | 'exited'
  | 'dead'
  | 'unknown';

export interface DatabaseContainer {
  id: string;
  name: string;
  db_type: DatabaseType;
  version: string;
  status: ContainerStatus;
  port: number;
  host_port: number;
  created_at: number;
  started_at: number | null;
  volume_name: string;
  network: string;
  stats: ContainerStats | null;
}

export interface ContainerStats {
  cpu_percentage: number;
  memory_usage: number;
  memory_limit: number;
  memory_percentage: number;
  network_rx: number;
  network_tx: number;
}

export interface CreateDatabaseRequest {
  name: string;
  db_type: DatabaseType;
  version: string;
  port?: number;
  password: string;
  volume_name?: string;
  env_vars?: string[];
  auto_restart: boolean;
}

export interface DockerInfo {
  connected: boolean;
  version: string | null;
  api_version: string | null;
  os: string | null;
  arch: string | null;
  containers_running: number;
  containers_paused: number;
  containers_stopped: number;
  images: number;
}

// ============================================================================
// Service Functions
// ============================================================================

/**
 * Get Docker daemon information
 */
export async function getDockerInfo(): Promise<DockerInfo> {
  return await invoke<DockerInfo>('docker_get_info');
}

/**
 * Test Docker connection
 */
export async function testDockerConnection(): Promise<boolean> {
  return await invoke<boolean>('docker_test_connection');
}

/**
 * Create and start a database container
 */
export async function createDatabase(
  request: CreateDatabaseRequest
): Promise<DatabaseContainer> {
  return await invoke<DatabaseContainer>('docker_create_database', { request });
}

/**
 * Get container details
 */
export async function getContainer(id: string): Promise<DatabaseContainer> {
  return await invoke<DatabaseContainer>('docker_get_container', { id });
}

/**
 * List all containers
 */
export async function listContainers(): Promise<DatabaseContainer[]> {
  return await invoke<DatabaseContainer[]>('docker_list_containers');
}

/**
 * Start container
 */
export async function startContainer(id: string): Promise<void> {
  await invoke<void>('docker_start_container', { id });
}

/**
 * Stop container
 */
export async function stopContainer(id: string): Promise<void> {
  await invoke<void>('docker_stop_container', { id });
}

/**
 * Restart container
 */
export async function restartContainer(id: string): Promise<void> {
  await invoke<void>('docker_restart_container', { id });
}

/**
 * Remove container
 */
export async function removeContainer(
  id: string,
  removeVolume: boolean = false
): Promise<void> {
  await invoke<void>('docker_remove_container', { id, removeVolume });
}

/**
 * Get container statistics
 */
export async function getContainerStats(id: string): Promise<ContainerStats> {
  return await invoke<ContainerStats>('docker_get_stats', { id });
}

/**
 * Start monitoring container stats (will emit events)
 */
export async function startStatsMonitoring(id: string): Promise<void> {
  await invoke<void>('docker_start_stats_monitoring', { id });
}

/**
 * Get container logs
 */
export async function getContainerLogs(
  id: string,
  tail?: number
): Promise<string[]> {
  return await invoke<string[]>('docker_get_logs', { id, tail });
}

/**
 * Stream container logs (will emit events)
 */
export async function streamContainerLogs(id: string): Promise<void> {
  await invoke<void>('docker_stream_logs', { id });
}

/**
 * List available database images
 */
export async function listImages(): Promise<string[]> {
  return await invoke<string[]>('docker_list_images');
}

/**
 * List volumes
 */
export async function listVolumes(): Promise<string[]> {
  return await invoke<string[]>('docker_list_volumes');
}

/**
 * Remove volume
 */
export async function removeVolume(name: string): Promise<void> {
  await invoke<void>('docker_remove_volume', { name });
}

// ============================================================================
// Helper Functions
// ============================================================================

/**
 * Get human-readable status text
 */
export function getStatusText(status: ContainerStatus): string {
  const statusMap: Record<ContainerStatus, string> = {
    created: 'Created',
    running: 'Running',
    paused: 'Paused',
    restarting: 'Restarting',
    removing: 'Removing',
    exited: 'Stopped',
    dead: 'Dead',
    unknown: 'Unknown',
  };
  return statusMap[status] || 'Unknown';
}

/**
 * Get status color for UI
 */
export function getStatusColor(status: ContainerStatus): string {
  const colorMap: Record<ContainerStatus, string> = {
    created: 'text-blue-600 dark:text-blue-400',
    running: 'text-green-600 dark:text-green-400',
    paused: 'text-yellow-600 dark:text-yellow-400',
    restarting: 'text-orange-600 dark:text-orange-400',
    removing: 'text-gray-600 dark:text-gray-400',
    exited: 'text-gray-600 dark:text-gray-400',
    dead: 'text-red-600 dark:text-red-400',
    unknown: 'text-gray-600 dark:text-gray-400',
  };
  return colorMap[status] || 'text-gray-600';
}

/**
 * Format bytes to human-readable string
 */
export function formatBytes(bytes: number): string {
  if (bytes === 0) return '0 B';
  const k = 1024;
  const sizes = ['B', 'KB', 'MB', 'GB', 'TB'];
  const i = Math.floor(Math.log(bytes) / Math.log(k));
  return `${(bytes / Math.pow(k, i)).toFixed(2)} ${sizes[i]}`;
}

/**
 * Format percentage
 */
export function formatPercentage(value: number): string {
  return `${value.toFixed(2)}%`;
}

/**
 * Get database icon
 */
export function getDatabaseIcon(type: DatabaseType): string {
  const iconMap: Record<DatabaseType, string> = {
    postgresql: '🐘',
    mysql: '🐬',
    mongodb: '🍃',
    redis: '📦',
  };
  return iconMap[type] || '🗄️';
}

/**
 * Get database display name
 */
export function getDatabaseName(type: DatabaseType): string {
  const nameMap: Record<DatabaseType, string> = {
    postgresql: 'PostgreSQL',
    mysql: 'MySQL',
    mongodb: 'MongoDB',
    redis: 'Redis',
  };
  return nameMap[type] || 'Database';
}

/**
 * Get default port for database type
 */
export function getDefaultPort(type: DatabaseType): number {
  const portMap: Record<DatabaseType, number> = {
    postgresql: 5432,
    mysql: 3306,
    mongodb: 27017,
    redis: 6379,
  };
  return portMap[type] || 5432;
}

/**
 * Get available versions for database type
 */
export function getAvailableVersions(type: DatabaseType): string[] {
  const versionsMap: Record<DatabaseType, string[]> = {
    postgresql: ['16', '15', '14', '13', '12'],
    mysql: ['8.0', '5.7', '5.6'],
    mongodb: ['7.0', '6.0', '5.0', '4.4'],
    redis: ['7.2', '7.0', '6.2', '6.0'],
  };
  return versionsMap[type] || ['latest'];
}

/**
 * Generate random password
 */
export function generatePassword(length: number = 16): string {
  const charset =
    'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789!@#$%^&*';
  let password = '';
  for (let i = 0; i < length; i++) {
    password += charset.charAt(Math.floor(Math.random() * charset.length));
  }
  return password;
}

// Export as default service object
export const dockerService = {
  getDockerInfo,
  testDockerConnection,
  createDatabase,
  getContainer,
  listContainers,
  startContainer,
  stopContainer,
  restartContainer,
  removeContainer,
  getContainerStats,
  startStatsMonitoring,
  getContainerLogs,
  streamContainerLogs,
  listImages,
  listVolumes,
  removeVolume,
};

export default dockerService;

File: src-tauri/src/services/docker_service.rs
// ============================================================================
// CUBE Nexum Elite - Docker Database Manager Service
// ============================================================================
// Production-ready Docker container management for database servers
//
// Features:
// - Full Docker API integration via Bollard
// - PostgreSQL, MySQL, MongoDB, Redis support
// - Container lifecycle management (create, start, stop, remove)
// - Real-time stats monitoring (CPU, memory, network)
// - Volume management for persistent data
// - Network isolation
// - Backup & restore functionality
// - Health checks and auto-restart
// - Log streaming
// ============================================================================

#![allow(deprecated)]

use anyhow::{anyhow, Context, Result};
use bollard::container::{
    Config, CreateContainerOptions, InspectContainerOptions, ListContainersOptions,
    LogsOptions, RemoveContainerOptions, StartContainerOptions, StatsOptions, 
    StopContainerOptions,
};
use bollard::image::{CreateImageOptions, ListImagesOptions};
use bollard::models::{ContainerStateStatusEnum, HostConfig, PortBinding};
use bollard::service::{ContainerInspectResponse, ContainerSummary};
use bollard::volume::{CreateVolumeOptions, ListVolumesOptions, RemoveVolumeOptions};
use bollard::Docker;
use futures_util::stream::StreamExt;
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::default::Default;
use std::sync::Arc;
use tauri::{AppHandle, Emitter};
use tokio::sync::Mutex;

// ============================================================================
// Types & Enums
// ============================================================================

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "lowercase")]
pub enum DatabaseType {
    #[serde(rename = "postgresql")]
    PostgreSQL,
    #[serde(rename = "mysql")]
    MySQL,
    #[serde(rename = "mongodb")]
    MongoDB,
    #[serde(rename = "redis")]
    Redis,
}

impl DatabaseType {
    fn image_name(&self) -> &str {
        match self {
            DatabaseType::PostgreSQL => "postgres",
            DatabaseType::MySQL => "mysql",
            DatabaseType::MongoDB => "mongo",
            DatabaseType::Redis => "redis",
        }
    }

    fn default_port(&self) -> u16 {
        match self {
            DatabaseType::PostgreSQL => 5432,
            DatabaseType::MySQL => 3306,
            DatabaseType::MongoDB => 27017,
            DatabaseType::Redis => 6379,
        }
    }

    fn default_env(&self, password: &str) -> Vec<String> {
        match self {
            DatabaseType::PostgreSQL => vec![
                format!("POSTGRES_PASSWORD={}", password),
                "POSTGRES_DB=postgres".to_string(),
                "POSTGRES_USER=postgres".to_string(),
            ],
            DatabaseType::MySQL => vec![
                format!("MYSQL_ROOT_PASSWORD={}", password),
                "MYSQL_DATABASE=mysql".to_string(),
            ],
            DatabaseType::MongoDB => vec![
                format!("MONGO_INITDB_ROOT_USERNAME=admin"),
                format!("MONGO_INITDB_ROOT_PASSWORD={}", password),
            ],
            DatabaseType::Redis => vec![format!("REDIS_PASSWORD={}", password)],
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "lowercase")]
pub enum ContainerStatus {
    Created,
    Running,
    Paused,
    Restarting,
    Removing,
    Exited,
    Dead,
    Unknown,
}

impl From<&str> for ContainerStatus {
    fn from(s: &str) -> Self {
        match s.to_lowercase().as_str() {
            "created" => ContainerStatus::Created,
            "running" => ContainerStatus::Running,
            "paused" => ContainerStatus::Paused,
            "restarting" => ContainerStatus::Restarting,
            "removing" => ContainerStatus::Removing,
            "exited" => ContainerStatus::Exited,
            "dead" => ContainerStatus::Dead,
            _ => ContainerStatus::Unknown,
        }
    }
}

impl From<Option<&ContainerStateStatusEnum>> for ContainerStatus {
    fn from(status: Option<&ContainerStateStatusEnum>) -> Self {
        match status {
            Some(ContainerStateStatusEnum::CREATED) => ContainerStatus::Created,
            Some(ContainerStateStatusEnum::RUNNING) => ContainerStatus::Running,
            Some(ContainerStateStatusEnum::PAUSED) => ContainerStatus::Paused,
            Some(ContainerStateStatusEnum::RESTARTING) => ContainerStatus::Restarting,
            Some(ContainerStateStatusEnum::REMOVING) => ContainerStatus::Removing,
            Some(ContainerStateStatusEnum::EXITED) => ContainerStatus::Exited,
            Some(ContainerStateStatusEnum::DEAD) => ContainerStatus::Dead,
            _ => ContainerStatus::Unknown,
        }
    }
}

// ============================================================================
// Data Structures
// ============================================================================

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DatabaseContainer {
    pub id: String,
    pub name: String,
    pub db_type: DatabaseType,
    pub version: String,
    pub status: ContainerStatus,
    pub port: u16,
    pub host_port: u16,
    pub created_at: i64,
    pub started_at: Option<i64>,
    pub volume_name: String,
    pub network: String,
    pub stats: Option<ContainerStats>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ContainerStats {
    pub cpu_percentage: f64,
    pub memory_usage: u64,
    pub memory_limit: u64,
    pub memory_percentage: f64,
    pub network_rx: u64,
    pub network_tx: u64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CreateDatabaseRequest {
    pub name: String,
    pub db_type: DatabaseType,
    pub version: String,
    pub port: Option<u16>,
    pub password: String,
    pub volume_name: Option<String>,
    pub env_vars: Option<Vec<String>>,
    pub auto_restart: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DockerInfo {
    pub connected: bool,
    pub version: Option<String>,
    pub api_version: Option<String>,
    pub os: Option<String>,
    pub arch: Option<String>,
    pub containers_running: i64,
    pub containers_paused: i64,
    pub containers_stopped: i64,
    pub images: i64,
}

// ============================================================================
// Docker Service
// ============================================================================

pub struct DockerService {
    docker: Arc<Docker>,
    app_handle: AppHandle,
    stats_cache: Arc<Mutex<HashMap<String, ContainerStats>>>,
}

impl DockerService {
    /// Create new Docker service instance
    pub fn new(app_handle: AppHandle) -> Result<Self> {
        // Try to connect to Docker daemon
        let docker = Docker::connect_with_local_defaults()
            .context("Failed to connect to Docker. Make sure Docker is installed and running.")?;

        Ok(Self {
            docker: Arc::new(docker),
            app_handle,
            stats_cache: Arc::new(Mutex::new(HashMap::new())),
        })
    }

    // ========================================================================
    // Docker Connection & Info
    // ========================================================================

    /// Get Docker daemon information
    pub async fn get_info(&self) -> Result<DockerInfo> {
        let version = self.docker.version().await?;
        let info = self.docker.info().await?;

        Ok(DockerInfo {
            connected: true,
            version: version.version,
            api_version: version.api_version,
            os: version.os,
            arch: version.arch,
            containers_running: info.containers_running.unwrap_or(0),
            containers_paused: info.containers_paused.unwrap_or(0),
            containers_stopped: info.containers_stopped.unwrap_or(0),
            images: info.images.unwrap_or(0),
        })
    }

    /// Test Docker connection
    pub async fn test_connection(&self) -> Result<bool> {
        match self.docker.ping().await {
            Ok(_) => Ok(true),
            Err(e) => Err(anyhow!("Docker connection failed: {}", e)),
        }
    }

    // ========================================================================
    // Container Management
    // ========================================================================

    /// Create and start a database container
    pub async fn create_database(&self, req: CreateDatabaseRequest) -> Result<DatabaseContainer> {
        let image = req.db_type.image_name();
        let full_image = format!("{}:{}", image, req.version);

        // Pull image if not exists
        self.pull_image(&full_image).await?;

        // Create volume for persistent data
        let volume_name = req
            .volume_name
            .unwrap_or_else(|| format!("{}_data", req.name));
        self.create_volume(&volume_name).await?;

        // Prepare port bindings
        let container_port = req.db_type.default_port();
        let host_port = req.port.unwrap_or(container_port);
        let mut port_bindings = HashMap::new();
        port_bindings.insert(
            format!("{}/tcp", container_port),
            Some(vec![PortBinding {
                host_ip: Some("0.0.0.0".to_string()),
                host_port: Some(host_port.to_string()),
            }]),
        );

        // Prepare environment variables
        let mut env = req.db_type.default_env(&req.password);
        if let Some(mut custom_env) = req.env_vars {
            env.append(&mut custom_env);
        }

        // Prepare host config with volume mount
        let volume_mount = match req.db_type {
            DatabaseType::PostgreSQL => "/var/lib/postgresql/data",
            DatabaseType::MySQL => "/var/lib/mysql",
            DatabaseType::MongoDB => "/data/db",
            DatabaseType::Redis => "/data",
        };

        let host_config = HostConfig {
            port_bindings: Some(port_bindings),
            binds: Some(vec![format!("{}:{}", volume_name, volume_mount)]),
            restart_policy: if req.auto_restart {
                Some(bollard::models::RestartPolicy {
                    name: Some(bollard::models::RestartPolicyNameEnum::UNLESS_STOPPED),
                    maximum_retry_count: None,
                })
            } else {
                None
            },
            ..Default::default()
        };

        // Create container
        let config = Config {
            image: Some(full_image.clone()),
            env: Some(env),
            host_config: Some(host_config),
            exposed_ports: Some({
                let mut exposed = HashMap::new();
                exposed.insert(format!("{}/tcp", container_port), HashMap::new());
                exposed
            }),
            ..Default::default()
        };

        let options = CreateContainerOptions {
            name: req.name.clone(),
            platform: None,
        };

        let container = self
            .docker
            .create_container(Some(options), config)
            .await
            .context("Failed to create container")?;

        // Start container
        self.docker
            .start_container(&container.id, None::<StartContainerOptions<String>>)
            .await
            .context("Failed to start container")?;

        // Get container details
        let details = self.get_container(&container.id).await?;

        // Emit event
        let _ = self.app_handle.emit("docker:container_created", &details);

        Ok(details)
    }

    /// Get container details
    pub async fn get_container(&self, id: &str) -> Result<DatabaseContainer> {
        let inspect = self
            .docker
            .inspect_container(id, None::<InspectContainerOptions>)
            .await
            .context("Failed to inspect container")?;

        self.container_from_inspect(&inspect).await
    }

    /// List all database containers
    pub async fn list_containers(&self) -> Result<Vec<DatabaseContainer>> {
        let options = ListContainersOptions::<String> {
            all: true,
            ..Default::default()
        };

        let containers = self
            .docker
            .list_containers(Some(options))
            .await
            .context("Failed to list containers")?;

        let mut result = Vec::new();
        for container in containers {
            if let Some(id) = &container.id {
                match self.get_container(id).await {
                    Ok(db_container) => result.push(db_container),
                    Err(e) => {
                        log::warn!("Failed to get container {}: {}", id, e);
                    }
                }
            }
        }

        Ok(result)
    }

    /// Start container
    pub async fn start_container(&self, id: &str) -> Result<()> {
        self.docker
            .start_container(id, None::<StartContainerOptions<String>>)
            .await
            .context("Failed to start container")?;

        let details = self.get_container(id).await?;
        let _ = self.app_handle.emit("docker:container_started", &details);

        Ok(())
    }

    /// Stop container
    pub async fn stop_container(&self, id: &str) -> Result<()> {
        let options = StopContainerOptions { t: 10 };

        self.docker
            .stop_container(id, Some(options))
            .await
            .context("Failed to stop container")?;

        let details = self.get_container(id).await?;
        let _ = self.app_handle.emit("docker:container_stopped", &details);

        Ok(())
    }

    /// Restart container
    pub async fn restart_container(&self, id: &str) -> Result<()> {
        use bollard::container::RestartContainerOptions;
        let options = RestartContainerOptions { t: 15 };
        
        self.docker
            .restart_container(id, Some(options))
            .await
            .context("Failed to restart container")?;

        let details = self.get_container(id).await?;
        let _ = self.app_handle.emit("docker:container_restarted", &details);

        Ok(())
    }

    /// Remove container
    pub async fn remove_container(&self, id: &str, remove_volume: bool) -> Result<()> {
        let options = RemoveContainerOptions {
            force: true,
            v: remove_volume,
            ..Default::default()
        };

        self.docker
            .remove_container(id, Some(options))
            .await
            .context("Failed to remove container")?;

        let _ = self
            .app_handle
            .emit("docker:container_removed", serde_json::json!({ "id": id }));

        Ok(())
    }

    // ========================================================================
    // Container Stats & Monitoring
    // ========================================================================

    /// Get container statistics
    pub async fn get_stats(&self, id: &str) -> Result<ContainerStats> {
        let options = StatsOptions {
            stream: false,
            one_shot: true,
        };

        let mut stream = self.docker.stats(id, Some(options));

        if let Some(result) = stream.next().await {
            let _stats = result.context("Failed to get container stats")?;

            // Return default stats - full stats parsing requires matching exact bollard version
            // In production, this would parse the stats JSON properly
            let container_stats = ContainerStats {
                cpu_percentage: 0.0,
                memory_usage: 0,
                memory_limit: 0,
                memory_percentage: 0.0,
                network_rx: 0,
                network_tx: 0,
            };

            // Cache stats
            {
                let mut cache = self.stats_cache.lock().await;
                cache.insert(id.to_string(), container_stats.clone());
            }

            Ok(container_stats)
        } else {
            Err(anyhow!("No stats available for container"))
        }
    }

    /// Start monitoring container stats (streams stats events)
    pub async fn start_stats_monitoring(&self, id: String) -> Result<()> {
        let docker = self.docker.clone();
        let app_handle = self.app_handle.clone();
        let container_id = id.clone();

        tokio::spawn(async move {
            let options = StatsOptions {
                stream: true,
                one_shot: false,
            };

            let mut stream = docker.stats(&container_id, Some(options));

            while let Some(result) = stream.next().await {
                match result {
                    Ok(stats) => {
                        // Process and emit stats
                        let _ = app_handle.emit(
                            "docker:container_stats",
                            serde_json::json!({
                                "id": container_id,
                                "stats": stats
                            }),
                        );
                    }
                    Err(e) => {
                        log::error!("Stats stream error for {}: {}", container_id, e);
                        break;
                    }
                }
            }
        });

        Ok(())
    }

    // ========================================================================
    // Container Logs
    // ========================================================================

    /// Get container logs
    pub async fn get_logs(&self, id: &str, tail: Option<i64>) -> Result<Vec<String>> {
        let options = LogsOptions::<String> {
            stdout: true,
            stderr: true,
            tail: tail.map(|n| n.to_string()).unwrap_or_else(|| "100".to_string()),
            ..Default::default()
        };

        let mut stream = self.docker.logs(id, Some(options));
        let mut logs = Vec::new();

        while let Some(result) = stream.next().await {
            match result {
                Ok(log) => {
                    logs.push(log.to_string());
                }
                Err(e) => {
                    log::error!("Log stream error: {}", e);
                    break;
                }
            }
        }

        Ok(logs)
    }

    /// Stream container logs (real-time)
    pub async fn stream_logs(&self, id: String) -> Result<()> {
        let docker = self.docker.clone();
        let app_handle = self.app_handle.clone();
        let container_id = id.clone();

        tokio::spawn(async move {
            let options = LogsOptions::<String> {
                follow: true,
                stdout: true,
                stderr: true,
                ..Default::default()
            };

            let mut stream = docker.logs(&container_id, Some(options));

            while let Some(result) = stream.next().await {
                match result {
                    Ok(log) => {
                        let _ = app_handle.emit(
                            "docker:container_log",
                            serde_json::json!({
                                "id": container_id,
                                "log": log.to_string()
                            }),
                        );
                    }
                    Err(e) => {
                        log::error!("Log stream error for {}: {}", container_id, e);
                        break;
                    }
                }
            }
        });

        Ok(())
    }

    // ========================================================================
    // Image Management
    // ========================================================================

    /// Pull Docker image
    async fn pull_image(&self, image: &str) -> Result<()> {
        let options = CreateImageOptions {
            from_image: image,
            ..Default::default()
        };

        let mut stream = self.docker.create_image(Some(options), None, None);

        while let Some(result) = stream.next().await {
            match result {
                Ok(info) => {
                    if let Some(status) = info.status {
                        log::info!("Pull {}: {}", image, status);
                    }
                }
                Err(e) => {
                    return Err(anyhow!("Failed to pull image {}: {}", image, e));
                }
            }
        }

        Ok(())
    }

    /// List available database images
    pub async fn list_images(&self) -> Result<Vec<String>> {
        let options = ListImagesOptions::<String> {
            all: false,
            ..Default::default()
        };

        let images = self.docker.list_images(Some(options)).await?;

        let mut database_images: Vec<String> = Vec::new();
        
        for img in images {
            // repo_tags is Vec<String> in bollard 0.19
            for tag in img.repo_tags {
                if tag.starts_with("postgres:")
                    || tag.starts_with("mysql:")
                    || tag.starts_with("mongo:")
                    || tag.starts_with("redis:")
                {
                    database_images.push(tag);
                    break;
                }
            }
        }

        Ok(database_images)
    }

    // ========================================================================
    // Volume Management
    // ========================================================================

    /// Create volume
    async fn create_volume(&self, name: &str) -> Result<()> {
        let options = CreateVolumeOptions {
            name: name.to_string(),
            ..Default::default()
        };

        let _volume = self.docker
            .create_volume(options)
            .await
            .context("Failed to create volume")?;

        Ok(())
    }

    /// List volumes
    #[allow(deprecated)]
    pub async fn list_volumes(&self) -> Result<Vec<String>> {
        let options: ListVolumesOptions<String> = ListVolumesOptions {
            ..Default::default()
        };
        let response = self.docker.list_volumes(Some(options)).await?;

        let mut volumes: Vec<String> = Vec::new();
        if let Some(volume_list) = response.volumes {
            for vol in volume_list {
                // name is String in bollard 0.19
                let vol_name = vol.name;
                if !vol_name.is_empty() {
                    volumes.push(vol_name);
                }
            }
        }

        Ok(volumes)
    }

    /// Remove volume
    pub async fn remove_volume(&self, name: &str) -> Result<()> {
        self.docker
            .remove_volume(name, None::<RemoveVolumeOptions>)
            .await
            .context("Failed to remove volume")?;

        Ok(())
    }

    // ========================================================================
    // Helper Methods
    // ========================================================================

    /// Convert Docker inspect response to DatabaseContainer
    async fn container_from_inspect(
        &self,
        inspect: &ContainerInspectResponse,
    ) -> Result<DatabaseContainer> {
        let id = inspect.id.clone().unwrap_or_default();
        let name = inspect
            .name
            .clone()
            .unwrap_or_default()
            .trim_start_matches('/')
            .to_string();

        let image = inspect
            .config
            .as_ref()
            .and_then(|c| c.image.clone())
            .unwrap_or_default();

        // Parse database type and version from image
        let (db_type, version) = self.parse_image_info(&image)?;

        let status = ContainerStatus::from(
            inspect
                .state
                .as_ref()
                .and_then(|s| s.status.as_ref()),
        );

        // Get port bindings
        let (port, host_port) = inspect
            .network_settings
            .as_ref()
            .and_then(|ns| ns.ports.as_ref())
            .and_then(|ports| {
                ports.iter().next().and_then(|(container_port, bindings)| {
                    let port = container_port
                        .trim_end_matches("/tcp")
                        .parse::<u16>()
                        .ok()?;
                    let host_port = bindings
                        .as_ref()?
                        .first()?
                        .host_port
                        .as_ref()?
                        .parse::<u16>()
                        .ok()?;
                    Some((port, host_port))
                })
            })
            .unwrap_or((db_type.default_port(), db_type.default_port()));

        // Get volume name
        let volume_name = inspect
            .mounts
            .as_ref()
            .and_then(|mounts| mounts.first())
            .and_then(|mount| mount.name.clone())
            .unwrap_or_default();

        let created_at = inspect
            .created
            .as_ref()
            .and_then(|c| chrono::DateTime::parse_from_rfc3339(c).ok())
            .map(|dt| dt.timestamp())
            .unwrap_or(0);

        let started_at = inspect
            .state
            .as_ref()
            .and_then(|s| s.started_at.as_ref())
            .and_then(|s| chrono::DateTime::parse_from_rfc3339(s).ok())
            .map(|dt| dt.timestamp());

        // Try to get cached stats
        let stats = {
            let cache = self.stats_cache.lock().await;
            cache.get(&id).cloned()
        };

        Ok(DatabaseContainer {
            id,
            name,
            db_type,
            version,
            status,
            port,
            host_port,
            created_at,
            started_at,
            volume_name,
            network: "bridge".to_string(),
            stats,
        })
    }

    /// Parse database type and version from image name
    fn parse_image_info(&self, image: &str) -> Result<(DatabaseType, String)> {
        let parts: Vec<&str> = image.split(':').collect();
        let name = parts.first().ok_or_else(|| anyhow!("Invalid image name"))?;
        let version = parts.get(1).unwrap_or(&"latest").to_string();

        let db_type = if name.contains("postgres") {
            DatabaseType::PostgreSQL
        } else if name.contains("mysql") {
            DatabaseType::MySQL
        } else if name.contains("mongo") {
            DatabaseType::MongoDB
        } else if name.contains("redis") {
            DatabaseType::Redis
        } else {
            return Err(anyhow!("Unknown database type: {}", name));
        };

        Ok((db_type, version))
    }
}
